gctscript/vm/manager.go

```diff
@@ -0,0 +1,123 @@
+
+	"github.com/thrasher-corp/gocryptotrader/common"
+	"github.com/thrasher-corp/gocryptotrader/engine/subsystem"
+	"github.com/thrasher-corp/gocryptotrader/log"
+)
+
+const gctscriptManagerName = "GCTScript"
+
+// GctScriptManager loads and runs GCT Tengo scripts
+type GctScriptManager struct {
+	config   *Config
+	started  int32
+	stopped  int32
+	shutdown chan struct{}
+	// Optional values to override stored config ('nil' if not overridden)
+	MaxVirtualMachines *uint8
+}
+
+// NewManager creates a new instance of script manager
+func NewManager(config *Config) (*GctScriptManager, error) {
+	if config == nil {
+		return nil, errors.New("config must be provided for script manager")
+	}
+	return &GctScriptManager{
+		config: config,
+	}, nil
+}
+
+// Started returns if gctscript manager subsystem is started
+func (g *GctScriptManager) Started() bool {
+	return atomic.LoadInt32(&g.started) == 1
+}
+
+// Start starts gctscript subsystem and creates shutdown channel
+func (g *GctScriptManager) Start(wg *sync.WaitGroup) (err error) {
+	if atomic.AddInt32(&g.started, 1) != 1 {
+		return fmt.Errorf("%s %s", gctscriptManagerName, subsystem.ErrSubSystemAlreadyStarted)
+	}
+
+	defer func() {
+		if err != nil {
+			atomic.CompareAndSwapInt32(&g.started, 1, 0)
+		}
+	}()
+	log.Debugln(log.Global, gctscriptManagerName, subsystem.MsgSubSystemStarting)
+
+	g.shutdown = make(chan struct{})
+	go g.run(wg)
+	return nil
+}
+
+// Stop stops gctscript subsystem along with all running Virtual Machines
+func (g *GctScriptManager) Stop() error {
+	if atomic.LoadInt32(&g.started) == 0 {
+		return fmt.Errorf("%s %s", gctscriptManagerName, subsystem.ErrSubSystemNotStarted)
+	}
+
+	if atomic.AddInt32(&g.stopped, 1) != 1 {
+		return fmt.Errorf("%s %s", gctscriptManagerName, subsystem.ErrSubSystemAlreadyStopped)
+	}
+
+	log.Debugln(log.GCTScriptMgr, gctscriptManagerName, subsystem.MsgSubSystemShuttingDown)
+	close(g.shutdown)
+	err := g.ShutdownAll()
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (g *GctScriptManager) run(wg *sync.WaitGroup) {
+	log.Debugln(log.Global, gctscriptManagerName, subsystem.MsgSubSystemStarted)
+
+	wg.Add(1)
+	SetDefaultScriptOutput()
+	g.autoLoad()
+	defer func() {
+		atomic.CompareAndSwapInt32(&g.stopped, 1, 0)
+		atomic.CompareAndSwapInt32(&g.started, 1, 0)
+		wg.Done()
+		log.Debugln(log.GCTScriptMgr, gctscriptManagerName, subsystem.MsgSubSystemShutdown)
+	}()
+
+	<-g.shutdown
+}
+
+func (g *GctScriptManager) autoLoad() {
+	for x := range g.config.AutoLoad {
+		temp := g.New()
+		if temp == nil {
+			log.Errorf(log.GCTScriptMgr, "Unable to create Virtual Machine, autoload failed for: %v",
+				g.config.AutoLoad[x])
+			continue
+		}
+		var name = g.config.AutoLoad[x]
+		if filepath.Ext(name) != common.GctExt {
+			name += common.GctExt
+		}
+		scriptPath := filepath.Join(ScriptPath, name)
+		err := temp.Load(scriptPath)
+		if err != nil {
+			log.Errorf(log.GCTScriptMgr, "%v failed to load: %v", filepath.Base(scriptPath), err)
```

Comment:
So this issue is outside of the scope of your PR, but it can be a "while you're here" 😄 completely optional and you don't have to fix this as it exists on `master` branch. 

An issue I've found is if there is an issue loading a script, for example that doesn't exist, it will still be loaded.
My config contains three scripts that don't exist:
```
"auto_load": [
   "hello",
   "mr",
   "rots"
  ],
```
Will keep 3 scripts loaded, even if they aren't found. If I then execute the following:
`gocryptotrader\cmd\gctcli> .\gctcli.exe script status`
I will get:
```
{
 "status": "3 of 10 virtual machines running",
 "scripts": [
  {
   "UUID": "fdc8f1f7-ffe3-4041-be0f-e6c997d2b6f9",
   "name": ".",
   "next_run": "0001-01-01 00:00:00 +0000 UTC"
  },
  {
   "UUID": "f4822f78-1cd7-4ae9-9995-e80f655df5b6",
   "name": ".",
   "next_run": "0001-01-01 00:00:00 +0000 UTC"
  },
  {
   "UUID": "ace9d470-3536-4b0d-a751-ac83741cffb7",
   "name": ".",
   "next_run": "0001-01-01 00:00:00 +0000 UTC"
  }
 ]
}
```

So to fix this one, what I'm suggesting is that when this error is hit on this line, that you also add
`err = temp.unregister()` so we can free up the virtual machines from error'd scripts that won't do anything

Reply:
Done